### PR TITLE
new windows module win_audit_policy_system

### PR DESCRIPTION
**Pull request by nwsparks**
##### SUMMARY

- New module for managing system wide audit policy on Windows. 
- Includes backup/restore capability
- Does not include user level audit management. It was in there initially but due to the complications it was causing I decided it would be best if it was moved to it's own module. I'll follow up at some point with another module for it. This is the reason for the module naming.
- No explicit modify in integration testing because "remove" is checking both things. Policies are never actually removed, what changes is just the type of events that are being recorded.
- There wasn't a short way of retrieving existing rules for the integration testing so the testing module was required. It's just returning a dictionary mapping the policy in question to the state of the logging for it (success, failure, both, no auditing). Assertions are then done against it.

##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME

win_audit_policy_system
##### ANSIBLE VERSION

```
[root@ip-10-5-10-122 ansible]# ansible --version
ansible 2.5.0 (win_audit_policy e8b881a977) last updated 2017/09/28 13:19:13 (GMT +000)
  config file = None
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /opt/ansible/lib/ansible
  executable location = /opt/ansible/bin/ansible
  python version = 2.7.12 (default, Sep  1 2016, 22:14:00) [GCC 4.8.3 20140911 (Red Hat 4.8.3-9)]

```

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

File: lib/ansible/modules/windows/win_audit_policy_system.ps1
Comment (jborean93) [inline lib/ansible/modules/windows/win_audit_policy_system.ps1]: There are 2 no-op branches in this if statement, why don't you just check the 2 if statement's that do something and just check the categories/sub categories if they are set.

File: lib/ansible/modules/windows/win_audit_policy_system.ps1
Comment (jborean93) [inline lib/ansible/modules/windows/win_audit_policy_system.ps1]: You should be setting `$result.changed = $true` straight after you run the auditpol executable.

**Review by jborean93 — COMMENTED** [review]
I still need to test this manually but I have a few thoughts on this

* I don't think the restore/backup options should be a function of this module. 

It doesn't really control the state of the audit policy but rather performs an action. It also is just a wrapper around calling the executable to run a particular task which can easily be done with `win_command`. Personally I think they should be removed to make the module more streamlined, and more purpose built to set the audit policy. Happy to be proven otherwise but I'll let other people comment on this.

* I am not a fan of calling `auditpol.exe` to perform these tasks, in the past calling executables has given us some grief especially when dealing with other languages.

Having a quick glance online there is no cmdlet/wmi or com object that we can easily call but we can use the Win32 API's through P/Invoke if necessary https://msdn.microsoft.com/en-us/library/aa375742(VS.85).aspx. P/Invoke does bring some more verbosity/complexity compared to cmdlets/wmi or com objects but it does make calling these tasks a lot simpler than doing string parsing on executable outputs.

I don't know how much experience you've had with calling the native Win32 API's from Powershell but that might be a better option if you want to take a go at it. There are a few modules currently where we do this, look at win_user_right if you want to see how it is done. Having a look at what is being performed it seems like there are a few calls that may need to be run but it might make managing policies easier in the future.

You may even want to consider using the `Run-Command` cmdlet in `Ansible.ModuleUtils.CommandUtil` to call the executable if you want to continue with using auditpol.exe. `Run-Command` can take in a string to execute and it returns a hashtable that contains the rc, stdout and stderr as separate objects. This makes it easier to validate if the command has run than when it is just called straight from Powershell.

Are you planning on creating separate modules for managing audit policies on the user level?

**Comment by nwsparks** [comment]
Thanks @jborean93 

I don't have issue with removing backup/restore. I do think there is some value in it being there in that it reminds people of the ability to perform those actions, but agree removing it makes things more streamlined. 

I did take a stab at one point with the win32 api, but wasn't able to get it to work...the effort I put in was pretty minimal though so I can try and revisit it.

I had planned to follow up with a separate module for the user specific settings. initially I had included it in this module, but it led to having a ton of mutually exclusive parameters due to how auditpol functions which made the module feel super clunky. There was a decent bit of weird behavior around this area of auditpol as well. If I'm able to get win32 api's working I might look at re-including it since the problems were mostly just stemming from auditpol.

I guess for now I'll try and look at the win32 api stuff and either try and get that implemented or if it still proves too troublesome I can follow up with a revision that uses run-command as you suggested.

**Comment by jborean93** [comment]
If you can't get the Win32 API working I don't want to hold this up, just let me know how you go and if you still want to continue the current track.

**Comment by nwsparks** [comment]
Ok. I don't really have any experience with .net or the api so it's been a bit of a struggle. I have portions of it done, but don't think I'd have a finished product any time soon.

I'll see if I can rework to use run-command in the next few days and recommit. I'll also be removing the backup/restore functions and fix the items you mentioned above.

**Comment by nwsparks** [comment]
Ok this should be good to go. LMK if any issues. ready_for_review

**Comment by nwsparks** [comment]
This one's been sitting for a bit. Did you guys want to see anymore changes?

File: lib/ansible/modules/windows/win_audit_policy_system.ps1
Comment (dagwieers) [inline lib/ansible/modules/windows/win_audit_policy_system.ps1]: Just being pedantic here, not a showstopper. But the syntax for copyright statement was changed to:
```powershell
# Copyright: (c) 2017, Noah Sparks <nsparks@outlook.com>
# Copyright: (c) 2017, Ansible Project
```

File: lib/ansible/modules/windows/win_audit_policy_system.ps1
Comment (dagwieers) [inline lib/ansible/modules/windows/win_audit_policy_system.ps1]: I personally don't like that style for comments. And it may trigger issues when we start doing style validation in the future ;-)

**Review by dagwieers — APPROVED** [review]
LGTM

File: lib/ansible/modules/windows/win_audit_policy_system.ps1
Comment (nwsparks) [reply lib/ansible/modules/windows/win_audit_policy_system.ps1]: thanks for checking it. You'd prefer just simple like this? 

`# comment here`

I like the fatter blocks to split sections up, but I don't mind changing it.

File: lib/ansible/modules/windows/win_audit_policy_system.ps1
Comment (nwsparks) [reply lib/ansible/modules/windows/win_audit_policy_system.ps1]: Np. I'll change since it looks like there's some new validation it failed for other items.

File: lib/ansible/modules/windows/win_audit_policy_system.ps1
Comment (dagwieers) [reply lib/ansible/modules/windows/win_audit_policy_system.ps1]: I'll leave it up to others to decided, as it is a personal opinion. (That's why I didn't ask for a change in this case).

Comment (nwsparks) [reply lib/ansible/modules/windows/win_audit_policy_system.ps1]: Ok gotcha

**Comment by jamesla** [comment]
Thank you for writing this hopefully it gets merged soon :-)

File: lib/ansible/modules/windows/win_audit_policy_system.py
Comment (jborean93) [inline lib/ansible/modules/windows/win_audit_policy_system.py]: Is this required anymore now that it is removed from the module.

File: lib/ansible/modules/windows/win_audit_policy_system.py
Comment (jborean93) [inline lib/ansible/modules/windows/win_audit_policy_system.py]: The more I think about this the more I think this should be a list type where you can specify some the options `success` and `failure`. You can use a combination of both (`success and failure`) or just an empty list (`none`) for the other options. We should also decide on a default, potentially `none` is the default if it isn't set.

**Review by jborean93 — COMMENTED** [review]
Only have 2 points so far but the rest looks great.

File: lib/ansible/modules/windows/win_audit_policy_system.py
Comment (nwsparks) [reply lib/ansible/modules/windows/win_audit_policy_system.py]: good catch

File: lib/ansible/modules/windows/win_audit_policy_system.py
Comment (nwsparks) [reply lib/ansible/modules/windows/win_audit_policy_system.py]: I think I did it this way to make the values match what auditpol outputs so I wouldn't have to do string manipulation. I can take a look though and see about changing it to a list.

I wonder if making it required instead of a default would be better here to force a selection. Otherwise, for a default, I think 'success' would make the most sense. I think people using this module would typically looking to be enable auditing rather than disable. Either way though you're right, it does need one of these. I'll give some thought about it...I guess there isn't much difference between the two really.

**Comment by nwsparks** [comment]
@jborean93 ok suggested changes have been made. docs updated, audit_type changed to list. I ended up doing failifempty for the audit_types. I think due to the nature of this, even if minor, making people take a little extra consideration for what they are adjusting would be good. If you really prefer the default though I can change it to -default success

File: lib/ansible/modules/windows/win_audit_policy_system.py
Comment (jborean93) [reply lib/ansible/modules/windows/win_audit_policy_system.py]: Sounds good to me

**Comment by jborean93** [comment]
LGTM, are you able to update the CHANGELOG once it is merged in.

rebuild_merge

**Comment by nwsparks** [comment]
thanks. will update changelog now.

**Pull request merged**
